### PR TITLE
feat: fly.Font

### PR DESCRIPTION
**Pull request by pudility**
The font class allows for a subset of glyphs from a font to be selected.

Example:
```JavaScript
  const ttf = await fetch('https://mdn.mozillademos.org/files/2468/VeraSeBd.ttf')
  const font = new fly.Font(await ttf.arrayBuffer())
  const subset = await font.subset(text)
```
`61 KB` -> `8 KB`

Todo:
 - [X] Write Tests
 - [X] Figure out where to put the typings file (fontmin)
- [x] Woff2/TTF support

### CODE REVIEW

**Comment by pudility** [comment]
Currently you need to add the following line to `node_modules/fontmin/package.json`
```JSON
  "typings": "../../src/fontmin.d.ts",
```

**Comment by pudility** [comment]
Looks like github needs to be added as a trusted IP on travis.

**Comment by jeromegn** [comment]
That Travis failure is unhelpful, doesn't seem to be much in terms of error message.

Are you running a "similar" environment locally? node 8.9 for instance :) This might help reproducing this error.

Not sure what more we can do though. I'll try running it locally if it keeps failing.

**Comment by pudility** [comment]
Yes, I have tried with that version on my machine, and I even tried messing with the travis config (on my own travis account).

I looked up the error and I think it has something to do with permissions.

File: src/bridge/fly/font.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    totalLength += arr.length;
+  }
+  let result = new Uint8Array(totalLength);
+  let offset = 0;
+  for (let arr of arrays) {
+    result.set(arr, offset);
+    offset += arr.length;
+  }
+  return <ArrayBuffer>result.buffer;
+}
```
Comment (jeromegn) [inline src/bridge/fly/font.ts]: This function can be replaced by `Buffer.concat(Array<Uint8Array>).buffer`

File: src/bridge/fly/font.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+/** @hidden */
+function toBuffer(ab:ArrayBuffer) {
+    var buf = new Buffer(ab.byteLength);
```
Comment (jeromegn) [inline src/bridge/fly/font.ts]: `new Buffer(size)` is deprecated, the proper way is `Buffer.alloc(size)`

Also, can this whole function be replaced with `Buffer.from(ab)`?

Comment (pudility) [reply src/bridge/fly/font.ts]: Yup - Thanks!

File: src/bridge/fly/font.ts (hunk 1 above)
Comment (pudility) [reply src/bridge/fly/font.ts]: Thanks, turns out it wasn't used in the first place so, I just got rid of it.

**Comment by jeromegn** [comment]
Tests also pass locally for me.

I tried installing them in a docker image to test on Linux, but it's proving more challenging than I expected.

I don't think this is a permissions issue like you mentioned. The error we see on travis is a generic error, we should've seen a stacktrace or at least an error message of some sort before the npm error stuff.

There seems to be something bombing while compiling the app.

I wonder if fontmin works well on linux? It seems to build fine for their own repo's Travis.

Unfortunately, when this happens and you can't reproduce locally, you need to add far more logging. A good first step might be to run it with LOG_LEVEL=silly on travis.

**Comment by pudility** [comment]
This afternoon I will try to run it on a linux VM. I think there might be an issue with this line: https://github.com/ecomfe/fontmin/blob/8f579ffffdfeb7a34ee1c23e8f873c9e49c6d7b9/lib/util.js#L21

**Comment by jeromegn** [comment]
@pudility oh yes, that would do it. I bet we need to install something on travis to make it work.

**Comment by mrkurt** [comment]
@pudility I'm pretty sure this PR isn't working. The spec filename has a typo, and doesn't pass locally when I rename it.

**Comment by jeromegn** [comment]
Yes, this PR assumes `Buffer` is available in the v8 environment. That's a node.js thing which doesn't work on fly.

I tried fixing the branch locally, but it always hung when it got to the `font.run` function. I gave up and went to work on something else ;)

**Pull request closed without merging**